README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# runway

<!-- badges: start -->

[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
<!-- badges: end -->

The goal of runway is to generate statistics and plots to calculate
discrimination, calibration, and decision curves for prediction models.

## Why is it called runway?

Because you can use it to visually compare models.

Sometimes your models look quite different.

![Models looking
different](https://i.pinimg.com/originals/2e/3d/14/2e3d14e6f382c6850685b5aaaff34fec.gif)

Other times your models look the same…

![Models looking
similar](https://pbs.twimg.com/media/Eg7RZKoXcAAhvKw?format=jpg&name=360x360)

## Installation

You can install `runway` from GitHub with:

``` r
remotes::install_github('ML4LHS/runway')
```

## Load the package

First, load the package.

``` r
library(runway)
```

## Sample datasets

Runway comes with two sample datasets.

``` r
data(single_model_dataset)
head(single_model_dataset)
#>   outcomes predictions
#> 1        0        0.36
#> 2        1        0.31
#> 3        0        0.39
#> 4        0        0.09
#> 5        0        0.44
#> 6        1        0.22

data(multi_model_dataset)
head(multi_model_dataset)
#>   outcomes predictions model_name
#> 1        0        0.26    Model 2
#> 2        1        0.28    Model 1
#> 3        0        0.56    Model 2
#> 4        0        0.27    Model 1
#> 5        0        0.31    Model 1
#> 6        0        0.42    Model 2
```

## Evaluating a single model

### Threshold-performance plot (single model)

``` r
threshperf_plot(single_model_dataset,
                outcome = 'outcomes',
                prediction = 'predictions')
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />

### Calibration plot with 10 bins (single model)

Note: 10 bins is the default.

``` r
cal_plot(single_model_dataset,
         outcome = 'outcomes', 
         prediction = 'predictions')
```

<img src="man/figures/README-unnamed-chunk-6-1.png" width="100%" />

### Calibration plot with 5 bins (single model)

``` r
cal_plot(single_model_dataset,
         outcome = 'outcomes', 
         prediction = 'predictions',
         n_bins = 5)
```

<img src="man/figures/README-unnamed-chunk-7-1.png" width="100%" />

### Calibration plot with 10 bins and loess curve (single model)

``` r
cal_plot(single_model_dataset,
         outcome = 'outcomes', 
         prediction = 'predictions',
         show_loess = TRUE)
```

<img src="man/figures/README-unnamed-chunk-8-1.png" width="100%" />

### Calibration plot with loess curve only (single model)

``` r
cal_plot(single_model_dataset,
         outcome = 'outcomes', 
         prediction = 'predictions',
         n_bins = 0,
         show_loess = TRUE)
```

<img src="man/figures/README-unnamed-chunk-9-1.png" width="100%" />

## Comparing multiple models

### Threshold-performance plot (multiple models)

``` r
threshperf_plot_multi(multi_model_dataset,
                      outcome = 'outcomes',
                      prediction = 'predictions',
                      model = 'model_name')
```

<img src="man/figures/README-unnamed-chunk-10-1.png" width="100%" />

### Calibration plot with 10 bins (multiple models)

Note: 10 bins is the default.

``` r
cal_plot_multi(multi_model_dataset,
         outcome = 'outcomes',
         prediction = 'predictions',
         model = 'model_name')
```

<img src="man/figures/README-unnamed-chunk-11-1.png" width="100%" />

## Calibration plot with 5 bins (multiple models)

``` r
cal_plot_multi(multi_model_dataset,
         outcome = 'outcomes',
         prediction = 'predictions',
         model = 'model_name',
         n_bins = 5)
```

<img src="man/figures/README-unnamed-chunk-12-1.png" width="100%" />

## Calibration plot with loess curve (multiple models)

Unlike single calibration plots, the choice of binned calibration and
loess calibration are mutually exclusive. To show LOESS curves, you must
set `show_loess` to `TRUE` *and* `n_bins` to `0`.

``` r
cal_plot_multi(multi_model_dataset,
         outcome = 'outcomes',
         prediction = 'predictions',
         model = 'model_name',
         n_bins = 0,
         show_loess = TRUE)
```

<img src="man/figures/README-unnamed-chunk-13-1.png" width="100%" />

## ROC curve w/CI

``` r
roc_plot(single_model_dataset, 
         outcome = 'outcomes', 
         prediction = 'predictions',
         ci = TRUE, 
         plot_title = 'Single ROC curve w/CI ribbon')
```

<img src="man/figures/README-unnamed-chunk-14-1.png" width="100%" />

## Multiple ROC curves w/CI ribbons

``` r
roc_plot_multi(multi_model_dataset, 
         outcome = 'outcomes', 
         prediction = 'predictions', 
         model = 'model_name',
         ci = TRUE,
         plot_title = 'Multiple model ROC curves w/CI ribbons')
```

<img src="man/figures/README-unnamed-chunk-15-1.png" width="100%" />


NEWS.md
# runway 0.0.0.9000

* Added additional statistics available for output in `threshperf(statistics=)`. The sensitivity and specificity are calculated using the {yardstick} package. All other statistics are then calculated using these values and the prevalence of the outcome. Calculating the statistics with this method allows a user to pass the prevalence and the performance statistics can also be calculated with a case-control design. (@ddsjoberg, #12)


tests/testthat.R
library(testthat)
library(runway)

test_check("runway")


tests/testthat/test-threshperf.R
test_that("threshperf() works", {
  expect_equal(
    single_model_dataset %>%
      threshperf(
        outcome = 'outcomes',
        prediction = 'predictions',
        thresholds = 0.5,
        statistics = c("sens", "spec", "ppv", "npv", "test_pos_rate", "test_neg_rate",
                       "tp_rate", "tn_rate", "fn_rate", "fp_rate", "prevalence")
      ) %>%
      dplyr::select(.metric, .estimate) %>%
      tibble::deframe(),
    c(sens = mean(single_model_dataset$predictions[single_model_dataset$outcomes == 1] >= 0.5),
      spec = mean(single_model_dataset$predictions[single_model_dataset$outcomes == 0] < 0.5),
      prevalence = mean(single_model_dataset$outcomes),
      tp_rate = mean(single_model_dataset$predictions >= 0.5 & single_model_dataset$outcomes == 1),
      fp_rate = mean(single_model_dataset$predictions >= 0.5 & single_model_dataset$outcomes == 0),
      tn_rate = mean(single_model_dataset$predictions < 0.5 & single_model_dataset$outcomes == 0),
      fn_rate = mean(single_model_dataset$predictions < 0.5 & single_model_dataset$outcomes == 1),
      test_pos_rate = mean(single_model_dataset$predictions >= 0.5),
      test_neg_rate = mean(single_model_dataset$predictions < 0.5),
      ppv = mean(single_model_dataset$outcomes[single_model_dataset$predictions >= 0.5]),
      npv = mean(!single_model_dataset$outcomes[single_model_dataset$predictions < 0.5])
    )
  )
})


R/helpers.R
expand_preds <- function (.data, threshold, inc = NULL) {
  threshold <- unique(threshold)
  nth <- length(threshold)
  n_data <- nrow(.data)
  if (!is.null(inc))
    .data <- dplyr::select(.data, all_of(inc))
  .data <- .data[rep(1:nrow(.data), times = nth), ]
  .data$.threshold <- rep(threshold, each = n_data)
  .data
}

recode_data <- function (obs, prob, threshold) {
  lvl <- levels(obs)
  if (getOption("yardstick.event_first", default = TRUE)) {
    pred <- ifelse(prob >= threshold, lvl[1], lvl[2])
  }
  else {
    pred <- ifelse(prob >= threshold, lvl[2], lvl[1])
  }
  factor(pred, levels = lvl)
}

two_class = function (...) {
  mets <- yardstick::metric_set(yardstick::sens,
                                yardstick::spec,
                                yardstick::j_index,
                                yardstick::accuracy,
                                yardstick::bal_accuracy,
                                yardstick::ppv,
                                yardstick::npv)
  mets(...)
}


R/cal_plots.R
#' Generate a single calibration plot with error bars showing 95 percent
#' confidence intervals
#'
#' This code builds off of code written by Darren Dahly, PhD in this blog post:
#' \url{https://darrendahly.github.io/post/homr/}.
#'
#' @param df The df as a data.frame.
#' @param outcome A character string containing the name of the column
#'   containing the outcomes (expressed as 0/1s).
#' @param prediction A character string containing the name of the column
#'   containing the predictions.
#' @param n_bins Number of bins. Defaults to 10. Set to 0 to hide binned
#'   calibration.
#' @param show_loess Whether to show loess smoothed calibration estimates.
#'   Defaults to FALSE. For \code{cal_plot()}, you can display both a binned
#'   calibration plot and a loess curve. In contrast, for
#'   \code{cal_plot_multi()}, these options are mutually exclusive.
#' @param plot_title A character string containing the title for the resulting
#'   plot.
#' @return A ggplot containing the calibration plot
#' @examples
#' data(single_model_dataset)
#' cal_plot(single_model_dataset, outcome = 'outcomes', prediction = 'predictions', n_bins = 5)
#' @export
cal_plot <- function(df, outcome, prediction, n_bins = 10, show_loess = FALSE, plot_title = '', ...){

  # The calibration plot
  g1 <- dplyr::mutate(df, bin = dplyr::ntile(!!parse_expr(prediction), n_bins)) %>%
    # Bin prediction into n_bins
    dplyr::group_by(bin) %>%
    dplyr::mutate(n = dplyr::n(), # Get ests and CIs
           bin_pred = mean(!!parse_expr(prediction), na.rm = TRUE),
           bin_prob = mean(as.numeric(as.character(!!parse_expr(outcome))), na.rm = TRUE),
           se = sqrt((bin_prob * (1 - bin_prob)) / n),
           ul = bin_prob + 1.96 * se,
           ll = bin_prob - 1.96 * se) %>%
    dplyr::mutate_at(dplyr::vars(ul, ll), . %>% scales::oob_squish(range = c(0,1))) %>%
    dplyr::ungroup() %>%
    ggplot2::ggplot() +
    ggplot2::scale_y_continuous(limits = c(0, 1), breaks = seq(0, 1, by = 0.1)) +
    ggplot2::scale_x_continuous(limits = c(0, 1), breaks = seq(0, 1, by = 0.1)) +
    ggplot2::geom_abline(linetype = 'dashed') # 45 degree line indicating perfect calibration

  if (n_bins > 0) {
    g1 = g1  +
      ggplot2::geom_point(ggplot2::aes(x = bin_pred, y = bin_prob, ymin = ll, ymax = ul),
                        size = 2, color = 'black') +
      ggplot2::geom_errorbar(ggplot2::aes(x = bin_pred, y = bin_prob, ymin = ll, ymax = ul),
                           size = 0.5, color = "black", width = 0.02)
  }
    # geom_smooth(method = "lm", se = FALSE, linetype = "dashed",
    #             color = "black", formula = y~-1 + x) +
    # straight line fit through estimates

  if (show_loess) {
    g1 = g1 +
    ggplot2::geom_smooth(ggplot2::aes(x = !!parse_expr(prediction), y = as.numeric(!!parse_expr(outcome))),
               color = "black", se = TRUE, method = "loess")
    # loess fit through estimates
  }

  g1 = g1 +
    ggplot2::xlab("Predicted Probability") +
    ggplot2::ylab("Observed Risk") +
    ggplot2::theme_minimal() +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::ggtitle(plot_title)

  # The distribution plot
  g2 <- ggplot2::ggplot(df, ggplot2::aes(x = !!parse_expr(prediction))) +
    ggplot2::geom_histogram(fill = "black", bins = 100) +
    ggplot2::scale_x_continuous(limits = c(0, 1), breaks = seq(0, 1, by = 0.1)) +
    ggplot2::xlab("") +
    ggplot2::ylab("") +
    ggplot2::theme_minimal() +
    ggeasy::easy_remove_y_axis() +
    ggeasy::easy_remove_x_axis() +
    ggplot2::theme_void() +
    ggplot2::theme(aspect.ratio = 0.1)

  # g2 / g1 + patchwork::plot_layout(heights = c(1,10))
  layout = c(patchwork::area(t = 1, b = 10, l = 1, r = 10),
             patchwork::area(t = 11, b = 12, l = 1, r = 10))

  g1 / g2

}


#' Generate multiple calibration plots with colored/shaded 95 percent confidence intervals
#'
#' This code builds off of code written by Darren Dahly, PhD in this blog post:
#' \url{https://darrendahly.github.io/post/homr/}.
#'
#' @param df The df as a data.frame.
#' @param outcome A character string containing the name of the column containing
#' the outcomes (expressed as 0/1s).
#' @param prediction A character string containing the name of the column containing
#' the predictions.
#' @param model A character string containing the name of the column containing
#' the model names
#' @param n_bins Number of bins. Defaults to 10. Set to 0 to hide binned calibration.
#' @param show_loess Whether to show loess smoothed calibration estimates. Defaults to FALSE.
#' For \code{cal_plot()}, you can display both a binned calibration plot and a loess curve. In
#' contrast, for \code{cal_plot_multi()}, these options are mutually exclusive. To display a loess
#' line, you must set \code{n_bins} to 0.
#' @param plot_title A character string containing the title for the resulting plot.
#' @return A ggplot containing the calibration plot
#' @examples
#' data(multi_model_dataset)
#' cal_plot_multi(multi_model_dataset, outcome = 'outcomes',model = 'model_name', prediction = 'predictions', n_bins = 5)
#' @export
cal_plot_multi <- function(df, outcome, prediction, model, n_bins = 10, show_loess = FALSE, plot_title = '', ...){

  if((n_bins > 0 && show_loess == TRUE) || (n_bins == 0 && show_loess == FALSE)) {
    stop('You must either set n_bins > 0 and show_loess to FALSE or set n_bins to 0 and show_loess to TRUE. Both cannot be displayed for cal_plot_multi()')
  }

  how_many_models = df[[model]] %>% unique() %>% length()

  # The calibration plot
  g1 <- df %>%
    dplyr::group_by(!!parse_expr(model)) %>%
    dplyr::mutate(bin = dplyr::ntile(!!parse_expr(prediction), n_bins)) %>%
    # Bin prediction
    dplyr::group_by(!!parse_expr(model), bin) %>%
    dplyr::mutate(n = dplyr::n(), # Get ests and CIs
           bin_pred = mean(!!parse_expr(prediction), na.rm = TRUE),
           bin_prob = mean(as.numeric(as.character(!!parse_expr(outcome))), na.rm = TRUE),
           se = sqrt((bin_prob * (1 - bin_prob)) / n),
           ul = bin_prob + 1.96 * se,
           ll = bin_prob - 1.96 * se) %>%
    dplyr::mutate_at(dplyr::vars(ul, ll), . %>% scales::oob_squish(range = c(0,1))) %>%
    dplyr::ungroup() %>%
    ggplot2::ggplot() +
    # geom_errorbar(size = 0.5, width = 0.02) +
    ggplot2::scale_y_continuous(limits = c(0, 1), breaks = seq(0, 1, by = 0.1)) +
    ggplot2::scale_x_continuous(limits = c(0, 1), breaks = seq(0, 1, by = 0.1)) +
    ggplot2::geom_abline(linetype = 'dashed') # 45 degree line indicating perfect calibration
    # geom_smooth(method = "lm", se = FALSE, linetype = "dashed",
    #            color = "black", formula = y~-1 + x) +
    # straight line fit through estimates
    # geom_smooth(ggplot2::aes(x = get(prediction), y = as.numeric(outcome)),
    #            color = "red", se = FALSE, method = "loess") +
    # loess fit through estimates
    # scale_color_viridis(discrete = TRUE, option = 'cividis', begin = 0.5) +
    # scale_fill_viridis(discrete = TRUE, option = 'cividis', begin = 0.5) +


    if (show_loess == FALSE && n_bins > 0) {
      g1 = g1 + ggplot2::aes(x = bin_pred,
                   y = bin_prob,
                   color = !!parse_expr(model),
                   fill = !!parse_expr(model)) +
        ggplot2::geom_ribbon(ggplot2::aes(ymin = ll,
                                          ymax = ul,),
                             alpha = 1/how_many_models) +
        ggplot2::geom_point(size = 2) +
        ggplot2::geom_line(size = 1, alpha = 1/how_many_models)
    }
    else if (show_loess == TRUE && n_bins == 0) {
      g1 = g1 +
        ggplot2::stat_smooth(ggplot2::aes(x = !!parse_expr(prediction), y = as.numeric(!!parse_expr(outcome)),
                                          color = !!parse_expr(model), fill = !!parse_expr(model)),
                            #              alpha = 1/how_many_models), # currently ignored by geom_smooth
                             se = TRUE, method = "loess")
      # loess fit through estimates
    }

  g1 = g1 +
    ggplot2::xlab("Predicted Probability") +
    ggplot2::ylab("Observed Risk") +
    ggplot2::scale_color_brewer(name = 'Models', palette = 'Set1') +
    ggplot2::scale_fill_brewer(name = 'Models', palette = 'Set1') +
    ggplot2::theme_minimal() +
    ggplot2::theme(aspect.ratio = 1) +
    ggplot2::ggtitle(plot_title)


  # The distribution plot
  g2 <- ggplot2::ggplot(df, ggplot2::aes(x = !!parse_expr(prediction))) +
    ggplot2::geom_density(alpha = 1/how_many_models, ggplot2::aes(fill = !!parse_expr(model), color = !!parse_expr(model))) +
    ggplot2::scale_x_continuous(limits = c(0, 1), breaks = seq(0, 1, by = 0.1)) +
    ggplot2::coord_fixed() +
    # scale_color_viridis(discrete = TRUE, option = 'cividis', begin = 0.5) +
    # scale_fill_viridis(discrete = TRUE, option = 'cividis', begin = 0.5) +
    ggplot2::xlab("") +
    ggplot2::ylab("") +
    ggplot2::scale_color_brewer(palette = 'Set1') +
    ggplot2::scale_fill_brewer(palette = 'Set1') +
    ggplot2::theme_minimal() +
    ggeasy::easy_remove_y_axis() +
    #  easy_remove_x_axis(what = c('ticks','line')) +
    ggeasy::easy_remove_legend(fill, color) +
    ggplot2::theme_void() +
    ggplot2::theme(aspect.ratio = 0.1)

  layout = c(patchwork::area(t = 1, b = 10, l = 1, r = 10),
             patchwork::area(t = 11, b = 12, l = 1, r = 10))

  g1 / g2

}



data-raw/generate_datasets.R
## code to prepare `single_model_dataset` dataset goes here

set.seed(1)

single_model_dataset =
  data.frame(
    outcomes = c(sample(c(0, 0, 1), 500, replace = TRUE),
                sample(c(1, 1, 0), 500, replace = TRUE)),
    predictions = c(sample(0:60/100, 500, replace = TRUE),
                    sample(40:100/100, 500, replace = TRUE)),
    stringsAsFactors = FALSE
  )

usethis::use_data(single_model_dataset, overwrite = TRUE)

set.seed(2)

multi_model_dataset =
  data.frame(
    outcomes = c(sample(c(0, 0, 1), 500, replace = TRUE),
                sample(c(1, 1, 0), 500, replace = TRUE)),
    predictions = c(sample(0:60/100, 500, replace = TRUE),
                    sample(40:100/100, 500, replace = TRUE)),
    model_name = c(sample(c('Model 1', 'Model 1', 'Model 2'), 500, replace = TRUE),
                   sample(c('Model 2', 'Model 2', 'Model 1'), 500, replace = TRUE)),
    stringsAsFactors = FALSE
  )

usethis::use_data(multi_model_dataset, overwrite = TRUE)


R/threshperf_plots.R
#' Calculates threshold-performance data
#'
#' This function is used by the \code{\link{threshperf_plot}} function. This is based on
#' the `threshold_perf()` function from the `probably` package (a part of `tidymodels`)
#' but expands the capability to allow for the range of metrics to include positive and
#' negative predictive value.

#' @param df The df as a data.frame.
#' @param outcome A character string containing the name of the column containing
#' the outcomes (expressed as 0/1s).
#' @param prediction A character string containing the name of the column containing
#' the predictions.
#' @param statistics Character vector of statistics to include.
#' Select among `c("sens", "spec", "ppv", "npv", "test_pos_rate", "test_neg_rate", "tp_rate", "tn_rate", "fn_rate", "fp_rate", "prevalence")`.
#' Default is `c("sens", "spec", "ppv", "npv")`.
#' @param prevalence Specify the prevalence of outcome for case-control studies.
#' Default is `NULL` and the prevalence will be estimated from `data`.
#' @param positive
#' @param thresholds
#' @return A data.frame containing the columns \code{.threshold}, \code{.metric},
#' \code{.estimator}, and \code{.estimate}
#' @examples
#' data(single_model_dataset)
#' threshperf(single_model_dataset, outcome = 'outcomes', prediction = 'predictions')
#' @export
threshperf <- function(df, outcome, prediction, positive = 'has_sepsis',
                       thresholds = NULL,
                       statistics = c("sens", "spec", "ppv", "npv"),
                       prevalence = NULL) {
  statistics <-
    match.arg(statistics,
              choices = c("sens", "spec", "ppv", "npv", "test_pos_rate", "test_neg_rate",
                          "tp_rate", "tn_rate", "fn_rate", "fp_rate", "prevalence"),
              several.ok = TRUE)

  if(is.null(thresholds)){
    thresholds = unique(c(0,sort(unique(df[[prediction]])), 1))
  }

  df <- dplyr::select(df, dplyr::all_of(c(outcome, prediction)))

  df <- stats::na.omit(df)

  df_orig <- df

  # IMPORTANT because order of levels matters to yardstick
  if (getOption('yardstick.event_first', default = TRUE)) {
    df[[outcome]] <- factor(df[[outcome]], levels = c(1,0))
  } else {
    df[[outcome]] <- factor(df[[outcome]], levels = c(0,1))
  }

  df <-
    df %>%
    expand_preds(threshold = thresholds,
                 inc = c(outcome, prediction))

  df <-
    df %>%
    dplyr::mutate(alt_pred = recode_data(!!rlang::parse_expr(outcome),
                                         !!rlang::parse_expr(prediction), .threshold))

  df <- df %>% dplyr::group_by(.threshold)

  df_metrics <-
    df %>%
    two_class(truth = !!rlang::parse_expr(outcome), estimate = alt_pred) %>%
    # adding other stats using the prevalence
    tidyr::pivot_wider(
      id_cols = dplyr::all_of(c(".threshold", ".estimator")),
      names_from = dplyr::all_of(".metric"),
      values_from = dplyr::all_of(".estimate")
    ) %>%
    dplyr::mutate(
      # TODO: Check prev calculation against the yardstick ordering!
      prevalence =
        .env$prevalence %||% unname((table(df$outcomes) / nrow(df))[2]),
      # all calculations use sens, spec, and prev to be able to handle
      # case-control data where true prev is not in data frame
      tp_rate = .data$sens * .data$prevalence,
      fp_rate = (1 - .data$spec) * (1 - .data$prevalence),
      tn_rate = .data$spec * (1 - .data$prevalence),
      fn_rate = (1 - .data$sens) * .data$prevalence,
      test_pos_rate = .data$tp_rate + .data$fp_rate,
      test_neg_rate = 1 - .data$test_pos_rate,
      ppv = .data$tp_rate / .data$test_pos_rate,
      npv = .data$tn_rate / .data$test_neg_rate
    ) %>%
    tidyr::pivot_longer(
      cols = -dplyr::all_of(c(".threshold", ".estimator")),
      names_to = ".metric",
      values_to = ".estimate"
    ) %>%
    dplyr::filter(.data$.metric %in% .env$statistics)

  suppressWarnings({df_metrics <-
    df_metrics %>%
    dplyr::group_by(.threshold) %>%
    dplyr::mutate(
      denom =
        dplyr::case_when(
          .data$.metric == 'sens' ~ sum(df_orig[[outcome]] == 1),
          .data$.metric == 'spec' ~ sum(df_orig[[outcome]] == 0),
          .data$.metric == 'ppv' ~ sum(df_orig[[prediction]] >= .threshold),
          .data$.metric == 'npv' ~ sum(df_orig[[prediction]] < .threshold),
          .data$.metric %in% c("tp_rate", "fp_rate", "tn_rate", "fn_rate",
                               "test_pos_rate", "test_neg_rate", "prevalence") ~ nrow(df_orig),

        )
    ) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(numer = round(.data$.estimate * .data$denom)) %>%
    stats::na.omit()})

  df_ci = Hmisc::binconf(x = df_metrics$numer, n = df_metrics$denom,
                         alpha = 0.05, method = 'wilson') %>%
    dplyr::as_tibble() %>%
    dplyr::rename(ll = Lower, ul = Upper) %>%
    dplyr::mutate_at(dplyr::vars(ul, ll), . %>% scales::oob_squish(range = c(0,1)))

  df_metrics = dplyr::bind_cols(df_metrics, df_ci)

  data.frame(df_metrics, check.names = FALSE, stringsAsFactors = FALSE)
}

#' Generate a threshold-performance plot for a single model
#'
#' @param df The df as a data.frame.
#' @param outcome A character string containing the name of the column
#'   containing the outcomes (expressed as 0/1s).
#' @param prediction A character string containing the name of the column
#'   containing the predictions.
#' @param show_denom Show the denominator (as a fraction of the maximum
#'   positives/negatives) for the positive predictive value (PPV) and negative
#'   predictive value (NPV).
#' @param plot_title A character string containing the title for the resulting
#'   plot.
#' @param xmin The minimum possible prediction. Defaults to 0.
#' @param xmax The maximum possible prediction. Defaults to 1.
#' @param pre_tp_geoms A ggplot geom_* (or a list of geoms) that should be drawn
#'   on the threshold-performance plot prior to rendering it.
#' @param pre_dist_geoms A ggplot geom_* (or a list of geoms) that should be
#'   drawn on the distribution plot prior to rendering it.
#' @param post_tp_geoms A ggplot geom_* (or a list of geoms) that should be
#'   drawn on the threshold-performance plot after rendering it.
#' @param post_dist_geoms A ggplot geom_* (or a list of geoms) that should be
#'   drawn on the distribution plot after rendering it.
#' @param heights A numeric vector of length 2 with ratio of heights of plots.
#'   Defaults to c(10, 1).
#' @param widths A numeric vector of length 3 with ratio of widths of plots.
#'   The first and third elements refer to padding. Defaults to c(1, 2, 1).
#' @param positive
#' @param thresholds
#' @return A ggplot containing the threshold-performance plot. The 95 percent
#'   confidence intervals are estimated using Wilson's interval from the
#'   \code{Hmisc} \code{\link[Hmisc]{binconf}} function.
#' @examples
#' data(single_model_dataset)
#' threshperf_plot(single_model_dataset, outcome = 'outcomes', prediction = 'predictions')
#' @export
threshperf_plot <- function(df, outcome, prediction, show_denom = TRUE, plot_title = '',
                            xmin = 0,
                            xmax = 1,
                            pre_tp_geoms = NULL,
                            pre_dist_geoms = NULL,
                            post_tp_geoms = NULL,
                            post_dist_geoms = NULL,
                            heights = c(10,1),
                            widths = c(1,2,1),positive = 'has_sepsis', thresholds = NULL) {
  tp_data = threshperf(df, outcome, prediction, thresholds)
  tp_data =
    tp_data %>%
    dplyr::group_by(.metric) %>%
    dplyr::mutate(denom_frac =
                    dplyr::if_else(.metric %in% c('ppv', 'npv'),
                                   denom/max(denom), NA_real_, NA_real_)) %>%
    dplyr::ungroup()

  tp_plot =
    tp_data %>%
    dplyr::mutate(.metric = dplyr::case_when(
      .metric == 'npv' ~ 'NPV',
      .metric == 'ppv' ~ 'PPV',
      .metric == 'spec' ~ 'Specificity',
      .metric == 'sens' ~ 'Sensitivity')) %>%
    dplyr::mutate(.metric = factor(.metric, levels = c('Sensitivity', 'Specificity', 'PPV', 'NPV'))) %>%
    dplyr::mutate_at(dplyr::vars(.estimate, ll, ul), . %>% {. * 100}) %>%
    ggplot2::ggplot(ggplot2::aes(x = .threshold,
                                 y = .estimate,
                                 ymin = ll,
                                 ymax = ul))

  if (!is.null(pre_tp_geoms)) {
    tp_plot =
      tp_plot +
      pre_tp_geoms
  }

  if (show_denom) {
    tp_plot =
      tp_plot +
      ggplot2::geom_ribbon(ggplot2::aes(ymin = 0, ymax = denom_frac*100), fill = 'grey', alpha = 1/3)
  }

  tp_plot = tp_plot +
    ggplot2::geom_ribbon(fill = 'grey') +
    ggplot2::geom_line(size = 1) +
    ggplot2::facet_grid(.metric~.) +
    ggplot2::theme_bw() +
    ggplot2::coord_cartesian(xlim=c(xmin,xmax)) +
    ggplot2::labs(x = 'Threshold', y = 'Performance (%)') +
    ggplot2::ggtitle(plot_title)

  if (!is.null(post_tp_geoms)) {
    tp_plot =
      tp_plot +
      post_tp_geoms
  }

  threshold_dist_plot =
    df %>%
    ggplot2::ggplot(ggplot2::aes(x=!!rlang::parse_expr(prediction)))

  if (!is.null(pre_dist_geoms)) {
    threshold_dist_plot =
      threshold_dist_plot +
      pre_dist_geoms
  }

  threshold_dist_plot =
    threshold_dist_plot +
    ggplot2::geom_histogram(fill = 'black', bins = 100) +
    ggplot2::coord_cartesian(xlim=c(xmin,xmax)) +
    ggplot2::theme_void()

  if (!is.null(post_dist_geoms)) {
    threshold_dist_plot =
      threshold_dist_plot +
      post_dist_geoms
  }

  patchwork::plot_spacer() +
    (tp_plot / threshold_dist_plot + patchwork::plot_layout(heights = heights)) +
    patchwork::plot_spacer() +
    patchwork::plot_layout(widths = widths)
}

#' Generate a threshold-performance plot for multiple models with colored/shaded
#' 95 percent confidence intervals
#'
#' @param df The df as a data.frame.
#' @param outcome A character string containing the name of the column
#'   containing the outcomes (expressed as 0/1s).
#' @param prediction A character string containing the name of the column
#'   containing the predictions.
#' @param model A character string containing the name of the column containing
#'   the model names
#' @param plot_title A character string containing the title for the resulting
#'   plot.
#' @param xmin The minimum possible prediction. Defaults to 0.
#' @param xmax The maximum possible prediction. Defaults to 1.
#' @param pre_tp_geoms A ggplot geom_* (or a list of geoms) that should be drawn
#'   on the threshold-performance plot prior to rendering it.
#' @param pre_dist_geoms A ggplot geom_* (or a list of geoms) that should be
#'   drawn on the distribution plot prior to rendering it.
#' @param post_tp_geoms A ggplot geom_* (or a list of geoms) that should be
#'   drawn on the threshold-performance plot after rendering it.
#' @param post_dist_geoms A ggplot geom_* (or a list of geoms) that should be
#'   drawn on the distribution plot after rendering it.
#' @param heights A numeric vector of length 2 with ratio of heights of plots.
#'   Defaults to c(10, 1).
#' @param widths A numeric vector of length 3 with ratio of widths of plots.
#'   The first and third elements refer to padding. Defaults to c(1, 2, 1).
#' @param thresholds
#' @return A ggplot containing the threshold-performance plot. The 95 percent
#'   confidence intervals are estimated using Wilson's interval from the
#'   \code{Hmisc} \code{\link[Hmisc]{binconf}} function.
#' @examples
#' data(multi_model_dataset)
#' threshperf_plot_multi(multi_model_dataset, outcome = 'outcomes', prediction = 'predictions', model = 'model_name')
#' @export
threshperf_plot_multi <- function(df, outcome, prediction, model, plot_title = '',
                                  xmin = 0,
                                  xmax = 1,
                                  pre_tp_geoms = NULL,
                                  pre_dist_geoms = NULL,
                                  post_tp_geoms = NULL,
                                  post_dist_geoms = NULL,
                                  heights = c(10,1),
                                  widths = c(1,2,1), thresholds = NULL) {

  how_many_models = df[[model]] %>% unique() %>% length()

  tp_data_list = list()
  for (model_name in unique(df[[model]])) {
    tp_data_list[[model_name]] <-
      threshperf(df[df[[model]] == model_name,],
                 outcome,
                 prediction, thresholds)
    tp_data_list[[model_name]][[model]] <- model_name
  }

  tp_data = dplyr::bind_rows(tp_data_list)

  tp_plot =
    tp_data %>%
    dplyr::mutate(.metric = dplyr::case_when(
      .metric == 'npv' ~ 'NPV',
      .metric == 'ppv' ~ 'PPV',
      .metric == 'spec' ~ 'Specificity',
      .metric == 'sens' ~ 'Sensitivity')) %>%
    dplyr::mutate(.metric = factor(.metric, levels = c('Sensitivity', 'Specificity', 'PPV', 'NPV'))) %>%
    dplyr::mutate_at(dplyr::vars(.estimate, ll, ul), . %>% {. * 100}) %>%
    ggplot2::ggplot(ggplot2::aes(x = .threshold,
                                 y = .estimate,
                                 ymin = ll,
                                 ymax = ul,
                                 color = !!rlang::parse_expr(model),
                                 fill = !!rlang::parse_expr(model)))


  if (!is.null(pre_tp_geoms)) {
    tp_plot =
      tp_plot +
      pre_tp_geoms
  }

  tp_plot = tp_plot +
    ggplot2::geom_ribbon(alpha = 1/how_many_models) +
    ggplot2::geom_line(size = 1) +
    ggplot2::facet_grid(.metric~.) +
    ggplot2::theme_bw() +
    ggplot2::coord_cartesian(xlim=c(xmin,xmax)) +
    ggplot2::labs(x = 'Threshold', y = 'Performance (%)') +
    ggplot2::scale_color_brewer(name = 'Models', palette = 'Set1') +
    ggplot2::scale_fill_brewer(name = 'Models', palette = 'Set1') +
    ggplot2::ggtitle(plot_title)

  if (!is.null(post_tp_geoms)) {
    tp_plot =
      tp_plot +
      post_tp_geoms
  }

  threshold_dist_plot <- ggplot2::ggplot(df, ggplot2::aes(x = !!rlang::parse_expr(prediction)))

  if (!is.null(pre_dist_geoms)) {
    threshold_dist_plot =
      threshold_dist_plot +
      pre_dist_geoms
  }

  threshold_dist_plot =
    threshold_dist_plot +
    ggplot2::geom_density(alpha = 1/how_many_models, ggplot2::aes(fill = !!rlang::parse_expr(model), color = !!rlang::parse_expr(model))) +
    ggplot2::scale_x_continuous(limits = c(xmin, xmax), breaks = seq(xmin, xmax, by = (xmax-xmin)/10)) +
    # scale_color_viridis(discrete = TRUE, option = 'cividis', begin = 0.5) +
    # scale_fill_viridis(discrete = TRUE, option = 'cividis', begin = 0.5) +
    ggplot2::xlab("") +
    ggplot2::ylab("") +
    ggplot2::scale_color_brewer(palette = 'Set1') +
    ggplot2::scale_fill_brewer(palette = 'Set1') +
    ggplot2::theme_minimal() +
    ggeasy::easy_remove_y_axis() +
    #  easy_remove_x_axis(what = c('ticks','line')) +
    ggeasy::easy_remove_legend(fill, color) +
    ggplot2::theme_void()

  if (!is.null(post_dist_geoms)) {
    threshold_dist_plot =
      threshold_dist_plot +
      post_dist_geoms
  }


  patchwork::plot_spacer() +
    (tp_plot / threshold_dist_plot + patchwork::plot_layout(heights = heights)) +
    patchwork::plot_spacer() +
    patchwork::plot_layout(widths = widths)
}
